wolfcrypt/src/port/Espressif/esp32_mp.c

```diff
@@ -132,75 +261,138 @@ static int esp_mp_hw_lock()
+     * with v5 release. Maybe it will be deprecated? */
     if (ret == 0) {
         periph_module_enable(PERIPH_RSA_MODULE);
 
         /* clear bit to enable hardware operation; (set to disable) */
-        DPORT_REG_CLR_BIT(DPORT_RSA_PD_CTRL_REG, DPORT_RSA_PD);
+        DPORT_REG_CLR_BIT(SYSTEM_RSA_PD_CTRL_REG, SYSTEM_RSA_MEM_PD);
     }
+#else
+    /* when unknown or not implmemted, assume there's no HW to lock */
 #endif
 
     /* reminder: wait until RSA_CLEAN_REG reads 1
-     *  see esp_mp_hw_wait_clean()
-     */
-
+    **   see esp_mp_hw_wait_clean() */
     ESP_LOGV(TAG, "leave esp_mp_hw_lock");
     return ret;
 }
 
 /*
-*   Release HW engine
+**  Release RSA HW engine
 */
-static void esp_mp_hw_unlock( void )
+static int esp_mp_hw_unlock( void )
 {
-#if CONFIG_IDF_TARGET_ESP32S3
-    /* Deactivate the RSA accelerator. See 20.3 of ESP32-S3 technical manual.
-     * periph_module_enable doesn't seem to be documented and in private folder
-     * with v5 release. Maybe it will be deprecated? */
-    DPORT_REG_SET_BIT(SYSTEM_RSA_PD_CTRL_REG, SYSTEM_RSA_MEM_PD);
-    periph_module_disable(PERIPH_RSA_MODULE);
-
+    int ret = MP_OKAY;
+    if (esp_mp_hw_islocked()) {
+
+#if defined(CONFIG_IDF_TARGET_ESP32)
+        /* set bit to disabled hardware operation; (clear to enable) */
+        DPORT_REG_SET_BIT(DPORT_RSA_PD_CTRL_REG, DPORT_RSA_PD);
+
+        /* Disable RSA hardware */
+        periph_module_disable(PERIPH_RSA_MODULE);
+#elif defined(CONFIG_IDF_TARGET_ESP32S3)
+        /* Deactivate the RSA accelerator. See 20.3 of ESP32-S3 technical manual.
+         * periph_module_enable doesn't seem to be documented and in private folder
+         * with v5 release. Maybe it will be deprecated? */
+        DPORT_REG_SET_BIT(SYSTEM_RSA_PD_CTRL_REG, SYSTEM_RSA_MEM_PD);
+        periph_module_disable(PERIPH_RSA_MODULE);
 #else
-    /* set bit to disabled hardware operation; (clear to enable)
-     */
-    DPORT_REG_SET_BIT(DPORT_RSA_PD_CTRL_REG, DPORT_RSA_PD);
-
-    /* Disable RSA hardware */
-    periph_module_disable(PERIPH_RSA_MODULE);
+        /* unknown platform, assume no HW to unlock  */
 #endif
+        /* unlock */
+#if defined(SINGLE_THREADED)
+        single_thread_locked = 0;
+#else
+        esp_CryptHwMutexUnLock(&mp_mutex);
+#endif /* SINGLE_THREADED */
+
+        ESP_LOGV(TAG, "esp_mp_hw_unlock");
+    }
+    else {
+        ESP_LOGW(TAG, "Warning: esp_mp_hw_unlock called when not locked.");
+    }
 
-    /* unlock */
-    esp_CryptHwMutexUnLock(&mp_mutex);
+    return ret;
 }
 
-/* this is based on an article by Cetin Kaya Koc,
- * A New Algorithm for Inversion: mod p^k, June 28 2017 */
+
+/* Only mulmod and mulexp_mod HW accelerator need montgomery math prep: M' */
+#if !defined(NO_WOLFSSL_ESP32_CRYPT_RSA_PRI_EXPTMOD) \
+      || \
+    !defined(NO_WOLFSSL_ESP32_CRYPT_RSA_PRI_MULMOD)
+
 static int esp_calc_Mdash(MATH_INT_T *M, word32 k, mp_digit* md)
 {
+    int ret = MP_OKAY;
+    ESP_LOGV(TAG, "\nBegin esp_calc_Mdash \n");
```

Comment:
potential log function called before the declaration of variable if USE_ALT_MPRIME is defined, we try to keep definition of variables at the top of the scope to avoid potential compiler errors/warnings

Reply:
ah, good catch. revised.